modules/ROOT/pages/dataweave-language-guide.adoc

```diff
@@ -1,75 +1,280 @@
+* `Value`, comprising any supported type, including another object, a function call, or an array.
+
+To declare an object, use braces (`{` and `}`) to enclose a comma-separated list of `Key:value` pairs:
+[source, dataweave, linenums]
 ----
 {
-  "title": "Java 8 in Action",
-  "author": "Mario Fusco",
-  "year": 2014
+  keyName @(keyAttributeName :"keyAttributeValue"): "value",
+  User @(role:"admin"): upper("max the mule"),
+  myArray: [1,2,3]
 }
 ----
 
-.Transform
-[source,dataweave,linenums]
+Objects also support the definition of conditional elements:
+[source, dataweave, linenums]
+----
+{
+ (userName: name) if !isBlank(name),
+ (userRole: role) if !isBlank(role)
+}
+----
+
+== Functional Programming Principles
+
+The functional programming paradigm enables developers to build applications by using pure function composition and immutable data. Because functional programming is declarative, functions in DataWeave are just expressions that return a value, instead of imperative statements that change program states.
+
+DataWeave follows the core principles of functional programming, including pure functions, first-class functions, high-order functions, function composition, lambda expressions, and immutable data.
+
+=== Pure Functions
+
+Pure functions in programming are similar to pure functions in mathematics in that, given a certain input, they always return the same output. Also, executing a function has no impact on other parts of a program, other than producing a result.
+All functions in DataWeave are pure; they are not affected by external values that could change the result of the execution.
+
+=== First-Class Functions
+
+The pure functions in DataWeave are also “first-class citizens” in that they support operations that are available to any other data type, including being stored in variables, passed as parameters, and returned as results.
+
+=== High-Order Functions
+
+High-order functions in programming also have the same concept as in mathematics. These functions take one or more functions as parameters and return a function as a result. The result of a high-order function can then be used as an input parameter in another function, enabling the concept of _function composition_.
+
+=== Function Composition
+
+Function composition is the process by which functions are combined to build complex operations. In this case, the result of executing one function serves as an input parameter for the next function. In DataWeave, all data transformations are a chain of expressions, functions composed to obtain a final result that outputs the data as needed.
+
+==== Example of Function Composition
+
+The following example uses the function `pluck` to map an object into an array, and then the resulting array serves as input for the function `filter`, which selects the elements in the array that have a `Role` equal to `ADMIN`. Also notice that during the mapping process, the function `upper` is used to transform one of the values:
```

Comment:
Suggestion: Split the first sentence into two. The second one would be "Then the resulting array..."